### DISCUSSION TITLE
Django Unicorn comparison

### DISCUSSION THREAD

**Discussion by gschurck** [tetra-framework/tetra#9] [Q&A] [answered]
Hello, very interesting project !
How does Tetra compare to Django Unicorn ?
It seems that the integration of Alpine.js is a big plus for client only actions.

**Comment by samwillis** [comment, marked as answer]
Thanks for taking a look!

As you probably know Alpine.js was created by the same developer as Livewire. Unicorn (which is incredible) is very much inspired by Livewire and I believe (I haven't used it) follows it closely. In the Laravel world you would use Alpine.js in combination with Livewire for, as you said, client slide only actions. One option with Unicorn is to use Alpine in combination with it, and that would work well.

What Tetra is doing differently is making Alpine.js a first class part of the system from the start, your public attributes on a Component become the reactive Alpine.js data model for the component on the client side. And then when a server side public method is called the updated client side data is available.

So I think Tetra is more "powerful" than Unicorn, in that is extends further into client side interactions. And (to copy from another thread):

- It also encapsulates all your Python, HTML, CSS and JavaScript related to a component in the same file. This is inspired by front end toolkits like Vue.js with it's "single file components". I'm not aware of any frameworks that have taken this approach and included the backend, it's the driving force behind the concept. I'm calling it "close proximity of related concerns" and its similar (although not exactly the same as) "locality of behaviour" (https://htmx.org/essays/locality-of-behaviour/).

- It provides "modern" JavaScript build tooling, your components can import any JavaScript and it's all nicely built and packaged for you using esbuild (even during development when using runserver). The intention is to make it easy to use modern JS libraries without having to use webpack or similar.

Unicorn, like Livewire, is aiming for zero front end JavaScript. Tetra is aiming for the same level of functionality with "zero JavaScript" but to have a route forward to build more complex front end components with JavaScript and reactive date.

So we are both inspired by some of the same things but Tetra is aiming to do more, particularly with front end JavaScript.

That said, Unicorn is stable and production ready now, I wouldn't advise Tetra for product work yet. The plan is to release a stable production ready v1 this summer.

**Reply by gschurck** [reply]
Thank you for your detailed answer.

I think Tetra is in a good spot. I am also interested by Django Unicorn, but I always think like if I want to build a serious project with user interactions, I will need to put at least a few JS at some point, because relying only on AJAX calls is not very optimal for some frontend interactions. So yes instead of integrating Alpine.js with Unicorn, Tetra seems to be the perfect fit for an easy to develop but complete frontend for Django without needing an API.

I am looking forward to see the evolution of this framework !

**Reply by nerdoc** [reply]
Having used Unicorn for a while (which is a GREAT project and does most things right), now finding Tetra is a big step forward. It just solves a few things that is missing in Unicorn: client side interactions with Alpine is the most notable one. Tetra seems to be faster too. the one-file approach is nice, but has drawbacks: having HTML/Django template code in one file helps you dealing with it better using an IDE: automatic indentation, code highlighting etc. does not work (outside of VSCode, here Tetra provides a plugin).

**Reply by gschurck** [reply]
Yes, I still use Unicorn currently as it's stable and has more features.
I also prefer the one-file approach like Svelte I find it more easy to work with, it's an annoying issue, a Pycharm plugin will be needed in the future.
But yeah Tetra looks promising.
